ldk/csharp/OliveHelpsLDK/OliveHelpsLDK/Filesystem/File.cs

```diff
@@ -0,0 +1,286 @@
+            {
+                Read = new FilesystemFileStreamRequest.Types.Read() { },
+            };
+            return WriteAndReceive(message,
+                response => response.Read,
+                read => read.Data.ToByteArray());
+        }
+
+        public Task Close()
+        {
+            var message = new FilesystemFileStreamRequest()
+            {
+                Close = new FilesystemFileStreamRequest.Types.Close() { },
+            };
+            CheckStatus();
+            var writeTask = WriteToStream(message);
+            Status = FileStatus.Closed;
+            Logger.Debug("File Stream Closed");
+            CompletionSource.SetResult(true);
+            return writeTask;
+        }
+
+        public Task ChangePermissions(int permission)
+        {
+            var message = new FilesystemFileStreamRequest()
+            {
+                Chmod = new FilesystemFileStreamRequest.Types.Chmod()
+                {
+                    Mode = (uint) permission
+                },
+            };
+            return WriteAndReceive(message,
+                response => response.Chmod,
+                (_) => true);
+        }
+
+        public Task ChangeOwnership(int userId, int groupId)
+        {
+            var message = new FilesystemFileStreamRequest()
+            {
+                Chown = new FilesystemFileStreamRequest.Types.Chown()
+                {
+                    Uid = userId,
+                    Gid = groupId
+                },
+            };
+            return WriteAndReceive(message,
+                response => response.Chown,
+                (_) => true);
+        }
+
+        public Task<FileInfo> FileInfo()
+        {
+            var message = new FilesystemFileStreamRequest()
+            {
+                Stat = new FilesystemFileStreamRequest.Types.Stat() { },
+            };
+            return WriteAndReceive(message,
+                response => response.Stat,
+                ConvertProtoFileInfo
+            );
+        }
+
+        public Task CompletionTask => CompletionSource.Task;
+
+
+        protected Proto.Session CreateSession()
+        {
+            return new Proto.Session
+            {
+                LoopID = Session.LoopId,
+                Token = Session.Token,
+            };
+        }
+
+        protected async Task<TOutput> WriteAndReceive<TResponse, TOutput>(Proto.FilesystemFileStreamRequest request,
+            Func<Proto.FilesystemFileStreamResponse, TResponse> responseReader,
+            Func<TResponse, TOutput> transformer,
+            [CallerMemberName] string caller = ""
+        )
+        {
+            CheckStatus();
+            await WriteToStream(request);
+            bool moveNext = false;
+            try
+            {
+                moveNext = await Stream.ResponseStream.MoveNext();
+            }
+            catch (Exception e)
+            {
+                Logger.Error("Duplex Response - Error on Move Next", e);
```

Comment:
This will be fine, in general, the interceptor will catch this too, but the extra logging and re-throw are generally fine.